airflow/sensors/external_task_sensor.py

```diff
@@ -223,21 +225,19 @@ def _handle_execution_date_fn(self, context):
         """
         This function is to handle backwards compatibility with how this operator was
         previously where it only passes the execution date, but also allow for the newer
-        implementation to pass all context through as well, to allow for more sophisticated
-        returns of dates to return.
-        Namely, this function check the number of arguments in the execution_date_fn
-        signature and if its 1, treat the legacy way, if it's 2, pass the context as
-        the 2nd argument, and if its more, throw an exception.
+        implementation to pass all context variables as keyword arguments, to allow
+        for more sophisticated returns of dates to return.
         """
-        num_fxn_params = self.execution_date_fn.__code__.co_argcount
-        if num_fxn_params == 1:
-            return self.execution_date_fn(context['execution_date'])
-        if num_fxn_params == 2:
-            return self.execution_date_fn(context['execution_date'], context)
-
-        raise AirflowException(
-            f'execution_date_fn passed {num_fxn_params} args but only allowed up to 2'
-        )
+        from airflow.utils.helpers import make_kwargs_callable
+
+        # Add "context" in the kwargs for backward compatibility (because context used to be
+        # an acceptable argument of execution_date_fn)
+        context["context"] = context
+        # Remove "execution_date" because it is already a mandatory positional argument
+        execution_date = context["execution_date"]
+        kwargs = {k: v for k, v in context.items() if k != "execution_date"}
```

Comment:
```suggestion
        execution_date = context.pop("execution_date")
        kwargs = {k: v for k, v in context.items()}
```

Reply:
Thanks for the suggestion. Thinking about it, doing so is going to change `context`, which is a `dict` passed in from `execute()` and may continue to be used after calling `execution_date_fn()`. It might work for now, but it seems to be a potential trap for future work (e.g. if we make a change that requires using `context["execution_date"]` in `poke()` or `execute()` after `execution_date_fn()`.

Reply:
Yes, removing things from context is bad